app/services/arcgis_api/token_keeper.rb

```diff
@@ -0,0 +1,236 @@
+module ArcgisApi
+  # Class for retrieve, refresh and manage caching of Arcgis API token.
+  # @attr_reader [connection] The underlying Faraday::Connection
+  # @attr_writer [prefetch_ttl] number of seconds to control timing of requesting a new token
+
+  # The token is saved in cache as a hash
+  # @example cache entry
+  # {
+  #   value: "XXXXXXXXXX",
+  #   sliding_expires_at: xxxxxxx,
+  #   expires_at: xxxxxxxxx,
+  # }
+  # expires_at: unix time that the token will expire, returned from
+  #   the Arcgis API generateToken call.
+  #
+  # sliding_expires_at: A time that the token does not actually
+  #   expire but used to control timing of requesting a new token before it expires.
+  #   It's initially set to expires_at - 3*prefetch_ttl. Only when sliding expiration enabled.
+  #
+  # If token synchronous fetching is disabled, a token will be fetched from cache directly
+  # or it will be a cache miss. Otherwise the thread will try to fetch/refresh
+  # the token on demand as needed.
+  #
+  # When sliding_expires_at<= current time <= sliding_expires_at + prefetch_ttl,
+  # the entry's sliding_expires_at time is updated to
+  # sliding_expires_at + prefetch_ttl and a new token is requested and saved to cache.
+  #
+  # When sliding_expires_at + prefetch_ttl < current time,
+  # a new token is requested and saved to cache.
+  #
+  # Optimistically, the token in cache will NOT expire
+  # when accessed by multi-threaded/distributed clients,
+  # since there is about prefetch_ttl - 2*prefetch_ttl
+  # length of time to generate a new API token.
+  #
+  class TokenKeeper
+    API_TOKEN_HOST = URI(IdentityConfig.store.arcgis_api_generate_token_url).host
+    API_TOKEN_CACHE_KEY = "arcgis_api_token:#{API_TOKEN_HOST}"
+    API_PREFETCH_TTL_SECONDS = 10
```

Comment:
Done

Reply:
@sheldon-b Given that the cache is shared by a variety of services, it would be good to ensure that cache keys (or at least their prefixes) are centrally visible & easily reconfigured in order to avoid and fix collisions.

Reply:
@NavaTim seems unnecessarily complex to me and adds a new class of subtle bugs that aren't possible if cache keys are constant/not configurable. Do you have any use cases when we would want to reconfigure a cache key?